processors/ratelimiter/ratelimiterprocessor.go

```diff
@@ -0,0 +1,120 @@
+package ratelimiter
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	pb_struct "github.com/envoyproxy/go-control-plane/envoy/extensions/common/ratelimit/v3"
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+	"go.opencensus.io/stats"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/metadata"
+)
+
+var _ component.TracesProcessor = (*processor)(nil)
+
+var (
+	droppedSpanCount = stats.Int64("dropped_span_count", "Number of spans dropped because of rate limiting", stats.UnitDimensionless)
+)
+
+func (p *processor) Start(_ context.Context, _ component.Host) error {
+	return nil
+}
+
+func (p *processor) Shutdown(_ context.Context) error {
+	return nil
+}
+
+func (p *processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: true}
+}
+
+type processor struct {
+	rateLimitServiceClient   pb.RateLimitServiceClient
+	domain                   string
+	domainSoftLimitThreshold uint32
+	logger                   *zap.Logger
+	tenantIDHeaderName       string
+	nextConsumer             consumer.Traces
+}
+
+const (
+	TenantID       = "tenant_id"
+	GlobalTenantID = "global_tenant_id"
+)
+
+// ProcessTraces implements processorhelper.ProcessTracesFunc
+func (p *processor) ConsumeTraces(ctx context.Context, traces ptrace.Traces) error {
+	tenantId, err := p.getTenantId(ctx)
+	if err != nil {
+		// If tenantId is missing, rate limiting not applicable.
+		return p.nextConsumer.ConsumeTraces(ctx, traces)
+	}
+	// two descriptors, one for tenant and other one for domain(cluster)
+	desc := make([]*pb_struct.RateLimitDescriptor, 2)
+	desc[0] = &pb_struct.RateLimitDescriptor{
+		Entries: []*pb_struct.RateLimitDescriptor_Entry{
+			{
+				Key:   TenantID,
+				Value: tenantId,
+			},
+		},
+	}
+	desc[1] = &pb_struct.RateLimitDescriptor{
+		Entries: []*pb_struct.RateLimitDescriptor_Entry{
+			{
+				Key:   TenantID,
+				Value: GlobalTenantID,
+			},
+		},
+	}
+	spanCount := uint32(traces.SpanCount())
+	response, err := p.rateLimitServiceClient.ShouldRateLimit(
+		ctx,
+		&pb.RateLimitRequest{
+			Domain:      p.domain,
+			Descriptors: desc,
+			HitsAddend:  spanCount,
+		})
+	if err != nil {
+		// Rate limit service call fails, spans will be forwarded as it is.
+		p.logger.Error("rate limit service call failed", zap.Error(err))
+		return p.nextConsumer.ConsumeTraces(ctx, traces)
+	}
+	descriptorStatuses := response.Statuses
+	if len(descriptorStatuses) != 0 {
+		if descriptorStatuses[1].GetCode() == pb.RateLimitResponse_OVER_LIMIT {
+			// If global rate limit exceeded drop all events
+			p.logger.Warn(fmt.Sprintf("dropping spans for tenant %s due to global rate limit exceeded, of spancount: %d", tenantId, traces.SpanCount()))
+			stats.Record(ctx, droppedSpanCount.M(int64(spanCount)))
```

Comment:
We need to associate the metric with the tenant id. See how the tenant id processor does it.

Reply:
Also can we distinguish this global rate limit drop from the soft rate limit one? by using 2 different metrics or labels.

Reply:
Updated, please check.